diff --git KMC-50eb19013e4eafa8034037ac43b7119c544377ed/Makefile KMC-latest/Makefile
index d286f4f..3c62fe3 100644
--- KMC-50eb19013e4eafa8034037ac43b7119c544377ed/Makefile
+++ KMC-latest/Makefile
@@ -13,15 +13,11 @@ OUT_BIN_DIR = bin
 OUT_INCLUDE_DIR = include
 
 ifeq ($(UNAME_S),Darwin)
-	CC = g++-11
-
-	CFLAGS	= -Wall -O3 -m64 -static-libgcc -static-libstdc++ -pthread -std=c++14
-	CLINK	= -lm -static-libgcc -static-libstdc++ -O3 -pthread -std=c++14
+	CFLAGS  = -Wall -O3 -m64 -pthread -std=c++14
+	CLINK   = -lm -O3 -pthread -std=c++14
 
 	PY_KMC_API_CFLAGS = -Wl,-undefined,dynamic_lookup -fPIC -Wall -shared -std=c++14 -O3
 else
-	CC 	= g++
-
 	CFLAGS	= -Wall -O3 -m64 -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
 	CLINK	= -lm -static -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
 
@@ -61,7 +57,7 @@ ifeq ($(UNAME_S),Darwin)
 	$(KMC_TOOLS_DIR)/libs/libz.1.2.5.dylib \
 	$(KMC_TOOLS_DIR)/libs/libbz2.1.0.5.dylib
 
-	LIB_KMC_CORE = $(OUT_BIN_DIR)/libkmc_core.mac.a
+	LIB_KMC_CORE = $(OUT_BIN_DIR)/libkmc_core.a
 else
 	RADULS_OBJS = \
 	$(KMC_MAIN_DIR)/raduls_sse2.o \
diff --git KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_CLI/kmc.cpp KMC-latest/kmc_CLI/kmc.cpp
index 15e579e..97b4af4 100644
--- KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_CLI/kmc.cpp
+++ KMC-latest/kmc_CLI/kmc.cpp
@@ -3,6 +3,7 @@
 #include <cstring>
 #include <iostream>
 #include <fstream>
+#include <random>
 #include <algorithm>
 #include <iomanip>
 using namespace std;
@@ -246,6 +247,9 @@ bool parse_parameters(int argc, char* argv[], Params& params)
 		stage2Params.SetOutputFileName(argv[i++]);
 	stage1Params.SetTmpPath(argv[i++]);
 
+	std::random_device rd;
+	std::mt19937 g(rd());
+
 	std::vector<std::string> input_file_names;	
 	if (input_file_name[0] != '@')
 		input_file_names.push_back(input_file_name);
@@ -264,7 +268,7 @@ bool parse_parameters(int argc, char* argv[], Params& params)
 				input_file_names.push_back(s);
 
 		in.close();
-		std::random_shuffle(input_file_names.begin(), input_file_names.end());
+		std::shuffle(input_file_names.begin(), input_file_names.end(), g);
 	}
 	stage1Params.SetInputFiles(input_file_names);
 
diff --git KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_api/kmer_defs.h KMC-latest/kmc_api/kmer_defs.h
index 90b0db4..4fe8aaf 100644
--- KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_api/kmer_defs.h
+++ KMC-latest/kmc_api/kmer_defs.h
@@ -34,7 +34,7 @@
 
 
 	#include <stdio.h>
-	#include <ext/algorithm>
+	#include <algorithm>
 	#include <iostream>
 
 #else
diff --git KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_core/defs.h KMC-latest/kmc_core/defs.h
index 75afc97..65322d3 100644
--- KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_core/defs.h
+++ KMC-latest/kmc_core/defs.h
@@ -109,8 +109,8 @@ using uint64 = uint64_t;
 #include <string.h>
 
 #include <stdio.h>
-#include <ext/algorithm>
-using __gnu_cxx::copy_n;
+#include <algorithm>
+//using __gnu_cxx::copy_n;
 
 #endif
 
diff --git KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_core/queues.h KMC-latest/kmc_core/queues.h
index 17a0465..4711f26 100644
--- KMC-50eb19013e4eafa8034037ac43b7119c544377ed/kmc_core/queues.h
+++ KMC-latest/kmc_core/queues.h
@@ -11,6 +11,9 @@
 #ifndef _QUEUES_H
 #define _QUEUES_H
 
+
+#include <algorithm>
+#include <random>
 #include "defs.h"
 #include <stdio.h>
 #include <tuple>
@@ -570,7 +573,10 @@ public:
 	}
 	void init_random()
 	{
-		lock_guard<mutex> lck(mtx);
+	  lock_guard<mutex> lck(mtx);
+	  std::random_device rd;
+ 	  std::mt19937 g(rd());
+
 		vector<pair<int32, int64>> bin_sizes;
 
 		for (auto& p : m)
@@ -589,7 +595,7 @@ public:
 		for (uint32 i = no_sort_end; i < bin_sizes.size(); ++i)
 			random_bins.push_back(bin_sizes[i].first);
 
-		random_shuffle(random_bins.begin(), random_bins.end());
+		shuffle(random_bins.begin(), random_bins.end(), g);
 
 		for (uint32 i = no_sort_start; i < no_sort_end; ++i)
 			random_bins.push_back(bin_sizes[i].first);
